api/pkg/s3/datatype/acl.go

```diff
@@ -0,0 +1,248 @@
+}
+
+type AccessControlPolicyResponse struct {
+	XMLName           xml.Name `xml:"AccessControlPolicy"`
+	Xmlns             string   `xml:"xmlns,attr,omitempty"`
+	ID                string   `xml:"Owner>ID"`
+	DisplayName       string   `xml:"Owner>DisplayName"`
+	AccessControlList []GrantResponse  `xml:"AccessControlList>Grant"`
+}
+
+type GrantResponse struct {
+	XMLName    xml.Name `xml:"Grant"`
+	Grantee    GranteeResponse  `xml:"Grantee"`
+	Permission string   `xml:"Permission"`
+}
+
+type GranteeResponse struct {
+	XMLName      xml.Name `xml:"Grantee"`
+	XmlnsXsi     string   `xml:"xmlns:xsi,attr"`
+	XsiType      string   `xml:"xsi:type,attr"`
+	URI          string   `xml:"URI,omitempty"`
+	ID           string   `xml:"ID,omitempty"`
+	DisplayName  string   `xml:"DisplayName,omitempty"`
+	EmailAddress string   `xml:"EmailAddress,omitempty"`
+}
+
+func IsValidCannedAcl(acl Acl) (err error) {
+	if !helper.StringInSlice(acl.CannedAcl, ValidCannedAcl) {
+		err = ErrInvalidCannedAcl
+		return
+	}
+	return
+}
+
+// the function will be deleted, because we will use AccessControlPolicy instead canned acl stored in hbase
+func GetCannedAclFromPolicy(policy AccessControlPolicy) (acl Acl, err error) {
+	aclOwner := Owner{ID: policy.ID, DisplayName: policy.DisplayName}
+	var canonUser bool
+	var group bool
+	for _, grant := range policy.AccessControlList {
+		log.Infoln("GetCannedAclFromPolicy")
+		switch grant.Grantee.XsiType {
+		case ACL_TYPE_CANON_USER:
+			if grant.Grantee.ID != aclOwner.ID {
+				log.Infoln("grant.Grantee.ID:", grant.Grantee.ID, "not equals aclOwner.ID:", aclOwner.ID)
+				return acl, ErrUnsupportedAcl
+			}
+			if grant.Permission != ACL_PERM_FULL_CONTROL {
+				log.Infoln("grant.Permission:", grant.Permission, "not equals", ACL_PERM_FULL_CONTROL)
+				return acl, ErrUnsupportedAcl
+			}
+			canonUser = true
+		case ACL_TYPE_GROUP:
+			if grant.Grantee.URI == ACL_GROUP_TYPE_ALL_USERS {
+				log.Infoln("grant.Grantee.URI is", ACL_GROUP_TYPE_ALL_USERS)
+				if grant.Permission != ACL_PERM_READ {
+					log.Infoln("grant.Permission:", grant.Permission, "not equals", ACL_PERM_READ)
+					return acl, ErrUnsupportedAcl
+				}
+				acl = Acl{CannedAcl: ValidCannedAcl[CANNEDACL_PUBLIC_READ]}
+				group = true
+			} else if grant.Grantee.URI == ACL_GROUP_TYPE_AUTHENTICATED_USERS {
+				log.Infoln("grant.Grantee.URI is", ACL_GROUP_TYPE_AUTHENTICATED_USERS)
+				if grant.Permission != ACL_PERM_READ {
+					log.Infoln("grant.Permission:", grant.Permission, "not equals", ACL_PERM_FULL_CONTROL)
+					return acl, ErrUnsupportedAcl
+				}
+				acl = Acl{CannedAcl: ValidCannedAcl[CANNEDACL_AUTHENTICATED_READ]}
+				group = true
+			} else {
+				log.Infoln("grant.Grantee.URI is invalid:", grant.Grantee.URI)
+				return acl, ErrUnsupportedAcl
+			}
+		default:
+			log.Infoln("grant.Grantee.XsiType is invalid:", grant.Grantee.XsiType)
+			return acl, ErrUnsupportedAcl
+		}
+	}
+
+	if !canonUser {
```

Comment:
partial to Canon eh ?? what about Nikon users ? ;-)

Reply:
@rhsakarpos , Sorry, what do you mean?  We only support canned acl,  and the function is used to extract canned acl from policy

Reply:
sorry, was joking. referring to Canon camera users and Nikon camera users, since your variable was named CanonUser :-)